src/transformers/sagemaker/trainer_sm.py

```diff
@@ -0,0 +1,177 @@
+# Copyright 2021 The HuggingFace Team. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from typing import Any, Dict, List, Optional, Tuple, Union
+
+import torch
+from torch import nn
+from torch.utils.data.dataset import Dataset
+from torch.utils.data.distributed import DistributedSampler
+
+from ..trainer import Trainer
+from ..trainer_pt_utils import (
+    DistributedLengthGroupedSampler,
+    SequentialDistributedSampler,
+    nested_detach,
+    nested_numpify,
+)
+from ..utils import logging
+from .training_args_sm import is_smdistributed_available
+
+
+logger = logging.get_logger(__name__)
+
+
+if is_smdistributed_available():
+    import smdistributed.modelparallel.torch as smp
+
+    @smp.step()
+    def forward_backward(model, inputs, gradient_accumulation_steps=1):
+        outputs = model(**inputs)
+        loss = outputs["loss"] if isinstance(outputs, dict) else outputs[0]
+        loss = loss / gradient_accumulation_steps
+        model.backward(loss)
+        return loss
+
+    @smp.step()
+    def forward_only(model, inputs):
+        return model(**inputs)
+
+    def smp_gather(tensor):
+        if isinstance(tensor, (list, tuple)):
+            return type(tensor)(smp_gather(t) for t in tensor)
+        elif isinstance(tensor, dict):
+            return type(tensor)({k: smp_gather(v) for k, v in tensor.items()})
+        elif not isinstance(tensor, torch.Tensor):
+            raise TypeError(
+                f"Can't gather the values of type {type(tensor)}, only of nested list/tuple/dicts of tensors."
+            )
+        all_tensors = smp.allgather(tensor, smp.CommGroup.DP_GROUP)
+        return torch.cat([t.cpu() for t in all_tensors], dim=0)
+
+    def nested_smp_concat(tensor):
```

Comment:
This method takes outputs from the `forward_only` method and (recursively) concatenates them in one tensor.

Reply:
The function is only used for outputs from our models, so list/tuple/dict of `StepOutput`. There might be a second level of nestedness (e.g. a tuple containing one tensor and a tuple of tensors) if the user set `output_hidden_states=True` or `output_attentions=True` but I doubt a user will get there.